build: setup `Pre-Commit`

keita-determined opened the pull request:
## Description

- [DET-8095]
- Setup [pre-commit](https://pre-commit.com/) for entire repo and setup `lint-staged` for web
<!---
Lead with the intended commit body in this description field. For breaking changes,
please include "BREAKING CHANGE:" at the beginning of your commit body.
At a minimum, this section should include a bracketed reference to the Jira ticket,
e.g. "[DET-1234]". When squash-and-merging, copy this directly into the description field.
-->

## How to Setup

- Run `make get-deps-pip`
- Run `make set-config-dev`
- Change file in `webui/react/src` folder
- Commit to run pre-commit

## Test Plan
- Check if all configurations are correct (especially pre-commit)
- Check if pre-commit is triggered correctly when specific files in web folder are staged and about to be committed
<!---
Describe the situations in which you've tested your change, and/or a screenshot as appropriate.
Reviewers may ask questions about this test plan to ensure adequate manual coverage of changes.
-->


## Commentary (optional)

<!---
Use this section of your description to add context to the PR. Could be for particularly
tricky bits of code that could use extra scrutiny, historical context useful for reviewers, etc.
You may intentionally leave this section blank and remove the title.
--->


## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.


<!---
## Title

Example title: "docs: tweak recommended "pip install" usage".

Specifically, this title should contain a type and a description
of the change being made:

User-facing change types:
- docs: docs-only change
- feat: new user-facing feature
- fix: bug fix
- perf: performance improvement

Internal change types:
- build: build system change (anything in a `Makefile`, mostly)
- chore: any internal change not covered by another type
- ci: anything that touches `.circleci`
- refactor: internal refactor
- style: style change
- test: new tests

See https://www.conventionalcommits.org/en/v1.0.0/ for background.

The first line should also:
- be at most 89 characters long
- contain a description that is at most 72 characters long
- not end with sentence-ending punctuation
- start (after the type) with a lowercase imperative ("add", "fix")
-->


[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DET-8093]: https://determinedai.atlassian.net/browse/DET-8093?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-8095]: https://determinedai.atlassian.net/browse/DET-8095?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ClaireNeveu commented:
Where is `pre-commit install` being called?

ClaireNeveu commented on requirements.txt:
```diff
@@ -22,6 +22,7 @@ flake8-tuple>=0.4.0
 isort==5.9.3
 mypy==0.910
 bump2version>=1.0.1
+pre-commit>=2.20
```
We should use `~=` here so we don't silently get a major version bump.

keita-determined commented on Makefile:
```diff
@@ -14,6 +14,7 @@ get-deps-%:
 get-deps-pip:
 	pip install torch==1.9.0
 	pip install -r requirements.txt
+	pre-commit init
```
Im not very familiar with makefile. I believe these commands are executed in serial, right?

ClaireNeveu replied:
We shouldn't do this in the dependency resolution step. You'll want to add a new step that operates on the pre-commit config file so that it only runs when that file changes and add that to the main process.

keita-determined commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,24 @@
+    hooks:
+      - id: web-js-lint-check
+        name: Web JS Lint Check
+        entry: bash -c 'make -j$(nproc) -C webui/react check-eslint check-prettier-js'
```
Added `nproc` to make pre-commit check faster.
On my mac (10 core CPU), 12s without `-j$(nproc)` option. 8s with `-j$(nproc)` option.
I saw an article that says j option with `nproc + 1` is the fastest but we dont need too much optimization and it was only 0.2s faster

dannysauer replied:
The n+1 is faster because it basically keeps a job in the queue. That's from the days when 2-4 CPUs was common, so with modern 10+ core systems, it might make sense to be more... Or maybe less with super fast disks. 🤷‍♂️

All that to say I'd support just leaving it at nprocs.

ioga replied:
make only runs two jobs here, `check-eslint` and `check-prettier-js` which have no deps. not much to parallelize. I suspect nproc > 2 doesn't do anything.

keita-determined replied:
I think so too, but the reason why I put `nproc` instead of a constant number is that if we add more commands in the future, we need to increment the number but I think we might forget to do it. 
I don't mind to switch to a constant number tho. Should I leave it or change it?

ioga replied:
keep nproc, it is good, I'd even support `nproc + 1` because it's canonical. 
I'm just pointing this out to highlight that benchmarking nproc=10 against nproc=11 may not be meaningful.

keita-determined commented:
I realized that `pre-commit` and `lint-staged` are overlapped features and `lint-staged` is unnecessary, so i removed `lint-staged`, and added more hooks in `pre-commit`. I think it's reasonable to do it, but if we don't agree with that, I can revert the commit

ioga reviewed (commented):
Q: does it still run when commit has no changes in `webui/react/src/`?

keita-determined commented:
> does it still run when commit has no changes in webui/react/src/?

`pre-commit` checks but if there's no change, `pre-commit` skip the checks like the following

<img width="406" alt="image" src="https://user-images.githubusercontent.com/109113616/192340850-8788ec50-04ce-4a2a-858b-551b8a6e60e1.png">

dannysauer commented:
You can theoretically force a re-check with `precommit run --all-files` or `precommit run --files ./some/path`, should that be something which is relevant for a specific directory.  In that case, typically one would likely have written their own check which knows, for example, two files should always be in sync and thus checks the unchanged file when the twin changes.  (coughs in version file)

dannysauer reviewed (commented):
Why is `webui/react/package-lock.json` changed in this PR? :)

Aside from that, I'm good with this.

keita-determined commented:
> Why is webui/react/package-lock.json changed in this PR? :)
>
> Aside from that, I'm good with this.

`npm uninstall lint-staged` didn't remove all changes. I'll fix it

ClaireNeveu reviewed (approved):
This looks good although we should note that DET-8093 is not actually covered by this PR

keita-determined commented:
> This looks good although we should note that DET-8093 is not actually covered by this PR

I'll untag the lint-staged

ClaireNeveu commented:
Actually, after reading the docs and testing it, pre-commit will only run against staged files. The only issue is that it doesn't seem to pass the staged files to the command, instead stashing the unstaged ones which makes the command take longer. I wonder if there's an alternate mode there.

keita-determined commented:
> Actually, after reading the docs and testing it, pre-commit will only run against staged files. The only issue is that it doesn't seem to pass the staged files to the command, instead stashing the unstaged ones which makes the command take longer. I wonder if there's an alternate mode there.

Yes, if we wanna run staged file, we need to use `pre-commit` offered libraries. I didn't wanna rely on that cuz some of them are made by individual and didn't seem reliable compared to official libs. 
I also considered that if there are some options to pass staged file in `pre-commit` but seems like there's no such things so far. We can achieve that using bash commands but that would make yaml file messy.
